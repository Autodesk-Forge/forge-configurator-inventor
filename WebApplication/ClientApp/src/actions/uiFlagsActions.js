/////////////////////////////////////////////////////////////////////
// Copyright (c) Autodesk, Inc. All rights reserved
// Written by Forge Design Automation team for Inventor
//
// Permission to use, copy, modify, and distribute this software in
// object code form for any purpose and without fee is hereby granted,
// provided that the above copyright notice appears in all copies and
// that both that copyright notice and the limited warranty and
// restricted rights notice below appear in all supporting
// documentation.
//
// AUTODESK PROVIDES THIS PROGRAM "AS IS" AND WITH ALL FAULTS.
// AUTODESK SPECIFICALLY DISCLAIMS ANY IMPLIED WARRANTY OF
// MERCHANTABILITY OR FITNESS FOR A PARTICULAR USE.  AUTODESK, INC.
// DOES NOT WARRANT THAT THE OPERATION OF THE PROGRAM WILL BE
// UNINTERRUPTED OR ERROR FREE.
/////////////////////////////////////////////////////////////////////

import repo from '../Repository';
import {addError} from './notificationActions';

export const actionTypes = {
    REJECT_PARAMETERS_EDITED_MESSAGE: 'REJECT_PARAMETERS_EDITED_MESSAGE',
    FETCH_SHOW_PARAMETERS_CHANGED: 'FETCH_SHOW_PARAMETERS_CHANGED',
    CLOSE_PARAMETERS_EDITED_MESSAGE: 'CLOSE_PARAMETERS_EDITED_MESSAGE',
    SHOW_MODAL_PROGRESS: 'SHOW_MODAL_PROGRESS',
    SHOW_UPDATE_FAILED: 'SHOW_UPDATE_FAILED',
    SHOW_LOGIN_FAILED: 'SHOW_LOGIN_FAILED',
    SHOW_DOWNLOAD_FAILED: 'SHOW_DOWNLOAD_FAILED',
    SET_ERROR_DATA: 'SET_ERROR_DATA',
    SHOW_DOWNLOAD_PROGRESS: 'SHOW_DOWNLOAD_PROGRESS',
    HIDE_DOWNLOAD_PROGRESS: 'HIDE_DOWNLOAD_PROGRESS',
    SET_DOWNLOAD_LINK: 'SET_DOWNLOAD_LINK',
    SHOW_UPLOAD_PACKAGE: 'SHOW_UPLOAD_PACKAGE',
    PACKAGE_FILE_EDITED: 'PACKAGE_FILE_EDITED',
    PACKAGE_ROOT_EDITED: 'PACKAGE_ROOT_EDITED',
    UPDATE_ACTIVE_TAB_INDEX: 'UPDATE_ACTIVE_TAB_INDEX',
    PROJECT_EXISTS: 'PROJECT_EXISTS',
    SHOW_DELETE_PROJECT: 'SHOW_DELETE_PROJECT',
    SET_PROJECT_CHECKED: 'SET_PROJECT_CHECKED',
    SET_CHECKED_PROJECTS: 'SET_CHECKED_PROJECTS',
    CLEAR_CHECKED_PROJECTS: 'CLEAR_CHECKED_PROJECTS',
    SHOW_DRAWING_PROGRESS: 'SHOW_DRAWING_PROGRESS',
    SHOW_ADOPT_WITH_PROPERTIES_PROGRESS: 'SHOW_ADOPT_WITH_PROPERTIES_PROGRESS',
    SHOW_ADOPT_WITH_PARAMS_FAILED: 'SHOW_ADOPT_WITH_PARAMS_FAILED',
    SET_DRAWING_URL: 'SET_DRAWING_URL',
    INVALIDATE_DRAWING: 'INVALIDATE_DRAWING',
    SET_STATS: 'SET_STATS',
    SET_REPORT_URL: 'SET_REPORT_URL',
    ACTIVE_DRAWING_UPDATED: 'ACTIVE_DRAWING_UPDATED',
    DRAWING_LIST_UPDATED: 'DRAWING_LIST_UPDATED'
};

export default actionTypes;

export const closeParametersEditedMessage = () => {
    return {
        type: actionTypes.CLOSE_PARAMETERS_EDITED_MESSAGE
    };
};

export const hideUpdateMessageBanner = (permanently) => async (dispatch) => {

    if (permanently === true) {
       const result = await repo.sendShowParametersChanged(false);
       dispatch(rejectParametersEditedMessage(!result));
    }

    dispatch(closeParametersEditedMessage());
};

export const rejectParametersEditedMessage = (show) => {
    return {
        type: actionTypes.REJECT_PARAMETERS_EDITED_MESSAGE,
        show
    };
};

export const fetchShowParametersChanged = () => async (dispatch) => {
    try {
        const showParametersChanged = await repo.loadShowParametersChanged();
        dispatch(rejectParametersEditedMessage(!showParametersChanged));
    } catch (error) {
        dispatch(addError('Failed to get information about "show changed parameters" . (' + error + ')'));
    }
};

export const showModalProgress = (visible) => {
    return {
        type: actionTypes.SHOW_MODAL_PROGRESS,
        visible
    };
};

export const showUpdateFailed = (visible) => {
    return {
        type: actionTypes.SHOW_UPDATE_FAILED,
        visible
    };
};

export const showLoginFailed = (visible) => {
    return {
        type: actionTypes.SHOW_LOGIN_FAILED,
        visible
    };
};

export const showDownloadFailed = (visible) => {
    return {
        type: actionTypes.SHOW_DOWNLOAD_FAILED,
        visible
    };
};

export const setErrorData = (errorData) => {
    return {
        type: actionTypes.SET_ERROR_DATA,
        errorData
    };
};

export const showDownloadProgress = (visible, title) => {
    return {
        type: actionTypes.SHOW_DOWNLOAD_PROGRESS,
        visible,
        title
    };
};

export const hideDownloadProgress = () => {
    return {
        type: actionTypes.HIDE_DOWNLOAD_PROGRESS
    };
};

export const showDrawingDownloadModalProgress = (visible) => {
    return {
        type: actionTypes.SHOW_DRAWING_DOWNLOAD_PROGRESS,
        visible
    };
};

export const setDownloadLink = (url) => {
    return {
        type: actionTypes.SET_DOWNLOAD_LINK,
        url
    };
};

export const showUploadPackage = (visible) => {
    return {
        type: actionTypes.SHOW_UPLOAD_PACKAGE,
        visible
    };
};

export const editPackageFile = (file, assemblies) => {
    return {
        type: actionTypes.PACKAGE_FILE_EDITED,
        file,
        assemblies
    };
};

export const editPackageRoot = (file) => {
    return {
        type: actionTypes.PACKAGE_ROOT_EDITED,
        file
    };
};

export const updateActiveTabIndex = (index) => {
    return {
        type: actionTypes.UPDATE_ACTIVE_TAB_INDEX,
        index
    };
};

export const setProjectAlreadyExists = (exists) => {
    return {
        type: actionTypes.PROJECT_EXISTS,
        exists
    };
};

export const showDeleteProject = (visible) => {
    return {
        type: actionTypes.SHOW_DELETE_PROJECT,
        visible
    };
};

export const setProjectChecked = (projectId, checked) => {
    return {
        type: actionTypes.SET_PROJECT_CHECKED,
        projectId,
        checked
    };
};

export const setCheckedProjects = (projects) => {
    return {
        type: actionTypes.SET_CHECKED_PROJECTS,
        projects
    };
};

export const clearCheckedProjects = () => {
    return {
        type: actionTypes.CLEAR_CHECKED_PROJECTS
    };
};

export const showDrawingExportProgress = (visible) => {
    return {
        type: actionTypes.SHOW_DRAWING_PROGRESS,
        visible
    };
};

export const showAdoptWithParametersProgress = (visible) => {
    return {
        type: actionTypes.SHOW_ADOPT_WITH_PROPERTIES_PROGRESS,
        visible
    };
};

export const showAdoptWithParamsFailed = (visible) => {
    return {
        type: actionTypes.SHOW_ADOPT_WITH_PARAMS_FAILED,
        visible
    };
};

export const setDrawingPdfUrl = (drawingKey, url) => {
    return {
        type: actionTypes.SET_DRAWING_URL,
        drawingKey,
        url
    };
};

export const invalidateDrawing = () => {
    return {
        type: actionTypes.INVALIDATE_DRAWING
    };
};

export const setStats = (stats, key) => {
    return {
        type: actionTypes.SET_STATS,
        stats,
        key
    };
};

export const setReportUrl = (reportUrl) => {
    return {
        type: actionTypes.SET_REPORT_URL,
        reportUrl
    };
};

export const updateActiveDrawing = activeDrawing => {
    return {
        type: actionTypes.ACTIVE_DRAWING_UPDATED,
        activeDrawing
    };
};

export const updateDrawingsList = drawingsList => {
    return {
        type: actionTypes.DRAWING_LIST_UPDATED,
        drawingsList
    };
};