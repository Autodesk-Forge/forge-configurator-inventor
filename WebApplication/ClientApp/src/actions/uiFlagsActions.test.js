/////////////////////////////////////////////////////////////////////
// Copyright (c) Autodesk, Inc. All rights reserved
// Written by Forge Design Automation team for Inventor
//
// Permission to use, copy, modify, and distribute this software in
// object code form for any purpose and without fee is hereby granted,
// provided that the above copyright notice appears in all copies and
// that both that copyright notice and the limited warranty and
// restricted rights notice below appear in all supporting
// documentation.
//
// AUTODESK PROVIDES THIS PROGRAM "AS IS" AND WITH ALL FAULTS.
// AUTODESK SPECIFICALLY DISCLAIMS ANY IMPLIED WARRANTY OF
// MERCHANTABILITY OR FITNESS FOR A PARTICULAR USE.  AUTODESK, INC.
// DOES NOT WARRANT THAT THE OPERATION OF THE PROGRAM WILL BE
// UNINTERRUPTED OR ERROR FREE.
/////////////////////////////////////////////////////////////////////

import Enzyme from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import * as uiFlagsActions from './uiFlagsActions';

// prepare mock for Repository module
jest.mock('../Repository');
import repoInstance from '../Repository';

Enzyme.configure({ adapter: new Adapter() });

// mock store
const middlewares = [thunk];
const mockStore = configureMockStore(middlewares);
const mockState = {
  uiFlags: {
  }
};
const store = mockStore(mockState);

describe('uiFlagsActions', () => {
    beforeEach(() => { // Runs before each test in the suite
        store.clearActions();
    });

    it('check closeParametersEditedMessage action', () => {
        store.dispatch(uiFlagsActions.closeParametersEditedMessage());
        expect(store.getActions()).toMatchSnapshot();
    });

    it('check hideUpdateMessageBanner action', async () => {
        const sendShowParametersChangedMock = repoInstance.sendShowParametersChanged;
        sendShowParametersChangedMock.mockReturnValue(true);

        await store.dispatch(uiFlagsActions.hideUpdateMessageBanner(true));
        expect(sendShowParametersChangedMock).toHaveBeenCalledTimes(1);
        expect(store.getActions()).toMatchSnapshot();
    });

    it ('check fetchShowParametersChanged action', async () => {
        const loadShowParametersChangedMock = repoInstance.loadShowParametersChanged;
        await store.dispatch(uiFlagsActions.fetchShowParametersChanged(true));
        expect(loadShowParametersChangedMock).toHaveBeenCalledTimes(1);
        expect(store.getActions()).toMatchSnapshot();
    });

    it('check rejectParametersEditedMessage action', () => {
        store.dispatch(uiFlagsActions.rejectParametersEditedMessage(true));
        expect(store.getActions()).toMatchSnapshot();
    });

    it('check showModalProgress action', () => {
        store.dispatch(uiFlagsActions.showModalProgress(true));
        expect(store.getActions()).toMatchSnapshot();
    });

    it('check showUpdateFailed action', () => {
        store.dispatch(uiFlagsActions.showUpdateFailed(true));
        expect(store.getActions()).toMatchSnapshot();
    });

    it('check showDownloadProgress action', () => {
        store.dispatch(uiFlagsActions.showDownloadProgress(true, "Some dialog title"));
        expect(store.getActions()).toMatchSnapshot();
    });

    it('check setRFALink action', () => {
        store.dispatch(uiFlagsActions.setDownloadLink("link"));
        expect(store.getActions()).toMatchSnapshot();
    });

    it('check showUploadPackage action', () => {
        store.dispatch(uiFlagsActions.showUploadPackage(true));
        expect(store.getActions()).toMatchSnapshot();
    });

    it('check editPackageFile action', () => {
        store.dispatch(uiFlagsActions.editPackageFile("file","assemblies"));
        expect(store.getActions()).toMatchSnapshot();
    });

    it('check editPackageRoot action', () => {
        store.dispatch(uiFlagsActions.editPackageRoot("root"));
        expect(store.getActions()).toMatchSnapshot();
    });

    it('check updateActiveTabIndex action', () => {
        store.dispatch(uiFlagsActions.updateActiveTabIndex(1));
        expect(store.getActions()).toMatchSnapshot();
    });

    it('check setProjectAlreadyExists action', () => {
        store.dispatch(uiFlagsActions.setProjectAlreadyExists(true));
        expect(store.getActions()).toMatchSnapshot();
    });

    it('check showDeleteProject action', () => {
        store.dispatch(uiFlagsActions.showDeleteProject(true));
        expect(store.getActions()).toMatchSnapshot();
    });

    it('check setCheckedProjects action', () => {
        store.dispatch(uiFlagsActions.setCheckedProjects(['1','2']));
        expect(store.getActions()).toMatchSnapshot();
    });

    it('check clearCheckedProjects action', () => {
        store.dispatch(uiFlagsActions.clearCheckedProjects());
        expect(store.getActions()).toMatchSnapshot();
    });
});
