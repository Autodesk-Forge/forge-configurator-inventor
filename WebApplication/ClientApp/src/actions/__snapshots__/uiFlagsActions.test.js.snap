// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`uiFlagsActions check clearCheckedProjects action 1`] = `
Array [
  Object {
    "type": "CLEAR_CHECKED_PROJECTS",
  },
]
`;

exports[`uiFlagsActions check closeParametersEditedMessage action 1`] = `
Array [
  Object {
    "type": "CLOSE_PARAMETERS_EDITED_MESSAGE",
  },
]
`;

exports[`uiFlagsActions check editPackageFile action 1`] = `
Array [
  Object {
    "assemblies": "assemblies",
    "file": "file",
    "type": "PACKAGE_FILE_EDITED",
  },
]
`;

exports[`uiFlagsActions check editPackageRoot action 1`] = `
Array [
  Object {
    "file": "root",
    "type": "PACKAGE_ROOT_EDITED",
  },
]
`;

exports[`uiFlagsActions check fetchShowParametersChanged action 1`] = `
Array [
  Object {
    "show": true,
    "type": "REJECT_PARAMETERS_EDITED_MESSAGE",
  },
]
`;

exports[`uiFlagsActions check hideUpdateMessageBanner action 1`] = `
Array [
  Object {
    "show": false,
    "type": "REJECT_PARAMETERS_EDITED_MESSAGE",
  },
  Object {
    "type": "CLOSE_PARAMETERS_EDITED_MESSAGE",
  },
]
`;

exports[`uiFlagsActions check rejectParametersEditedMessage action 1`] = `
Array [
  Object {
    "show": true,
    "type": "REJECT_PARAMETERS_EDITED_MESSAGE",
  },
]
`;

exports[`uiFlagsActions check setCheckedProjects action 1`] = `
Array [
  Object {
    "projects": Array [
      "1",
      "2",
    ],
    "type": "SET_CHECKED_PROJECTS",
  },
]
`;

exports[`uiFlagsActions check setProjectAlreadyExists action 1`] = `
Array [
  Object {
    "exists": true,
    "type": "PROJECT_EXISTS",
  },
]
`;

exports[`uiFlagsActions check setRFALink action 1`] = `
Array [
  Object {
    "type": "SET_DOWNLOAD_LINK",
    "url": "link",
  },
]
`;

exports[`uiFlagsActions check showDeleteProject action 1`] = `
Array [
  Object {
    "type": "SHOW_DELETE_PROJECT",
    "visible": true,
  },
]
`;

exports[`uiFlagsActions check showDownloadProgress action 1`] = `
Array [
  Object {
    "title": "Some dialog title",
    "type": "SHOW_DOWNLOAD_PROGRESS",
    "visible": true,
  },
]
`;

exports[`uiFlagsActions check showModalProgress action 1`] = `
Array [
  Object {
    "type": "SHOW_MODAL_PROGRESS",
    "visible": true,
  },
]
`;

exports[`uiFlagsActions check showUpdateFailed action 1`] = `
Array [
  Object {
    "type": "SHOW_UPDATE_FAILED",
    "visible": true,
  },
]
`;

exports[`uiFlagsActions check showUploadPackage action 1`] = `
Array [
  Object {
    "type": "SHOW_UPLOAD_PACKAGE",
    "visible": true,
  },
]
`;

exports[`uiFlagsActions check updateActiveTabIndex action 1`] = `
Array [
  Object {
    "index": 1,
    "type": "UPDATE_ACTIVE_TAB_INDEX",
  },
]
`;
